Use Full 32 Bits Of TP ADC Integral

aeoranday opened the pull request:
The TP ADC integral field was only being used for the low 16 bits. The changes in this PR separates the AVX segment of integral calculations into the low 16 bits and high 16 bits. Then, the finalized TP combines to allow usage of the full 32 bits.

This change removes the early closing of TPs that saturate the low 16 bits, and it does **not** shift this early closing to the full 32 bits. Any channel that is saturating at 32 bits should be considered problematic as it would need to send saturated WIBEth signals (14 bit) for ~134 ms.

For testing, I applied the following patch and ran `$DBT_AREA_ROOT/build/tpglibs/unittest/avx_generator_test`:
```
diff --git a/unittest/avx_generator_test.cxx b/unittest/avx_generator_test.cxx
index fe0c22a..bec905f 100644
--- a/unittest/avx_generator_test.cxx
+++ b/unittest/avx_generator_test.cxx
@@ -26,9 +26,12 @@ BOOST_AUTO_TEST_CASE(test_macro_overview)
 
   const int num_time_samples_per_frame = dunedaq::fddetdataformats::WIBEthFrame::s_time_samples_per_frame;
 
+  uint16_t signal = 0x3333u;
+  uint16_t signal_count = 5;
   for (int t_sample = 0; t_sample < num_time_samples_per_frame; t_sample++) {
     for (int chan = 0; chan < 64; chan++) {
-      frame.set_adc(chan, t_sample, (t_sample*100 + chan) % 500);
+      uint16_t val = t_sample % (signal_count + 1) == 0 ? 0 : signal;
+      frame.set_adc(chan, t_sample, val);
     }
   }
 
@@ -44,8 +47,8 @@ BOOST_AUTO_TEST_CASE(test_macro_overview)
       "AVXThresholdProcessor",
       {
         {"plane0", 200},
-        {"plane1", 300},
-        {"plane2", 445}
+        {"plane1", 200},
+        {"plane2", 200}
       }
     }
   };
@@ -61,9 +64,11 @@ BOOST_AUTO_TEST_CASE(test_macro_overview)
 
   for (auto tp : tps) {
     if (tp.adc_peak < min_peak) min_peak = tp.adc_peak;
+    fmt::print("TP ADC Integral == {}\n", tp.adc_integral);
+    BOOST_TEST(tp.adc_integral == (uint32_t)signal * (uint32_t)signal_count);
   }
 
-  BOOST_TEST(tp_count == 600);
+//  BOOST_TEST(tp_count == 600);
   BOOST_TEST(min_peak > 200);  // Truly it should depend on the plane, so this is naive.
 }

```

To be thorough, I varied the values of `signal` and `signal_count`. For the WIBEth format, `signal` must be less than or equal to `0x3FFF` and `signal_count` must be less than 64. I tested the following cases:
* High value integrals that need the new expansion. Mixture of short signals with high values, long signals with low values, and long signals with high values.
* Low value integrals that would not need the new expansion. Mix of short signals with low values and long signals with low values.
* Integrals that are exact multiples of `0xFFFF` (`uint16_t` saturation). Set signal at `0x3333` and `signal_count` as multiples of 5.
I saw all cases succeeding with the expected integral values for each.

This was **not** tested on a live system.

aeoranday committed:
Add High-Low 16 Bits of Integral.

Combine later to get the full 32-bit integral.

aeoranday committed:
Add `exact` Edge Case

Added the edge case when the signal pushes the lo integral exactly to 0xFFFF and not over.

roland-sipos commented on src/AVXPipeline.cpp:
```diff
@@ -47,22 +61,24 @@ std::vector<dunedaq::trgdataformats::TriggerPrimitive>
     if (tp_tot[i] == 0) continue;  // Don't track non-TPs.
     dunedaq::trgdataformats::TriggerPrimitive tp;
-    tp.adc_integral        = tp_integral[i];
+    tp.adc_integral        = tp_integral_lo[i] + ((uint32_t)tp_integral_hi[i] << 16);
```
Please use C++ style casting, also cast up the first var.

roland-sipos reviewed (changes requested):
Comment inline, and please run a micro-bench to see the extra instructions' overhead.

aeoranday committed:
Use More Descriptive Variable Name.

`tmp` is too vague. Changed to `over` since this is the overflown integral.

aeoranday committed:
Use C++ Function-Style Cast.

aeoranday commented:
Micro-bench saw a drastic increase in ns per call as reported by `gprof`. The original single call `_mm256_adds_epu16` varied with 0.4 - 0.7 ns per call. The separate high-low integral instead saw 12 - 14 ns per call.

This was also run live with warm CRPs (actual use of the high 16 bits is unlikely). There each TPG core had an additional 3-4% usage compared to the 16-bit saturation integral.

aeoranday committed:
Change Saturation/Overflow Calculation Order.

aeoranday committed:
Use Const Registers.

There are a few spots that a 1s or max valued register is being used as a base. This commit makes a private, const data member so they aren't built repetitively.

aeoranday commented:
Performance now sees a minimal loss online: 0.087 ns / call to 0.103 ns / call (~18% increase). The benefit of having 32 bit ADC integration seems to outweigh this performance loss.

roland-sipos reviewed (approved):
:no_mouth:
